web/modules/custom/dpl_go/src/DplGoServiceProvider.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\dpl_go;
+
+use Drupal\Core\DependencyInjection\ContainerBuilder;
+use Drupal\Core\DependencyInjection\ServiceProviderBase;
+
+/**
+ * Ensures `www.` prefix is stripped from cookie_domain.
+ *
+ * On sites whose primary domain starts with `www`, Drupal sets the cookie
+ * domain to `.www.<site>` per default. This means that the cookie isn't shared
+ * with `www.go.<site>` and the login doesn't work. In these cases we set it
+ * explicitly without the `.www` prefix.
+ */
+class DplGoServiceProvider extends ServiceProviderBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function alter(ContainerBuilder $container): void {
+    $parameter = 'session.storage.options';
+
+    if (!$container->getParameterBag()->has($parameter)) {
+      return;
+    }
+
+    /** @var array<string, string> $cookieSettings */
+    $cookieSettings = $container->getParameter($parameter);
+
+    // The cookie_domain shouldn't be set as this comes directly from the
+    // services.yml file, but for safety, use any configured value.
+    if (isset($cookieSettings['cookie_domain'])) {
+      $cookieDomain = $cookieSettings['cookie_domain'];
+    }
+    else {
+      // We can't use the service from `dpl_lagoon` for this, as obviously we
+      // can't use the container while building it.
+      $route = getenv('LAGOON_ROUTE');
+
+      if (!$route) {
+        // Without a route, we can't do much.
+        return;
+      }
+
+      // Split off the scheme.
+      $parts = explode('//', $route, 2);
```

Comment:
Well done!
I think using parse_url would be more elegant here.
Then it should be possible to write:
```php
['host' => $host] = parse_url($route);